### PR TITLE
mgr/cephadm: Allow customizing mgr/cephadm/lsmcli_blink_lights_cmd per host

### PR DESCRIPTION
**Pull request by votdev**
* Rename key name from `lsmcli_blink_lights_cmd` to `blink_device_light_cmd`
* Refactor TemplateMgr::render() method to use the Ceph common behavior how to name store/module option keys. The old implementation required a key like `mgr/cephadm/services_nfs_ganesha.conf` instead of `mgr/cephadm/services/nfs/ganesha.conf` or `mgr/cephadm/mgr0_blink_device_light_cmd` instead of `mgr/cephadm/mgr0/blink_device_light_cmd`.

Fixes: https://tracker.ceph.com/issues/48041

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by votdev** [comment]
jenkins test make check

**Comment by jmolmo** [comment]
I think that the way to solve this is probably not the right one.

1. You are providing a way to the user to "customize" the command used to switch on/off a light in one host, but:
- this "unknown" command must work with the parameters (all/part) we are using with the "lsmcli" command. If the command needs other parameters this is not going to work.
- You have no way to know if this is going to work or even cause any kind of problem in the client cluster. 
- In any case, if this goes forward. You must indicate that there is not guaranteed that this is going to work and can have unexpected results. And ...  Where is the documentation that allows the user to know how to change the template?

2. We are using the "lsmcli" command. This command belongs to the libStorageMgmt library:
And this library has been tested with the current set of hardware:
https://libstorage.github.io/libstoragemgmt-doc/doc/user_guide.html#supported-hardware

I think we should include this previous warning (about hardware supported) in the documentation, in:
https://docs.ceph.com/en/latest/rados/operations/devices/#identifying-physical-devices

In this way we are saying explicitly that this is only going to work with a set of hardware, and any other hardware is not supported. (avoiding this kind of dangerous modification )

3. IMHO, if we need to support different commands to switch on/off lights we need to have very clear what is the hardware that is going to be used, the command to use and the exact library/package which provides the command. And obviously we need to test that it works with cephadm. Otherwise, the most probable result are unexpected things, and those of this things can be bad....

**Comment by jmolmo** [comment]
Another thing...
This introduces a way to execute any "ad-hoc" command in the cluster hosts. From the point of view of security also it is not a good  idea.

File: src/pybind/mgr/cephadm/tests/test_cephadm.py [hunk 1]
```diff
@@ -645,12 +645,24 @@ def test_blink_device_light(self, _run_cephadm, on_bool, fault_ident, cephadm_mo
     def test_blink_device_light_custom(self, _run_cephadm, cephadm_module):
         _run_cephadm.return_value = '{}', '', 0
         with with_host(cephadm_module, 'test'):
-            cephadm_module.set_store('lsmcli_blink_lights_cmd', 'echo hello')
+            cephadm_module.set_store('blink_device_light_cmd', 'echo hello')
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/tests/test_cephadm.py:648]: Users can replace "echo hello" by "cd / && rm -rf" without problems.
Demonstration of how dangerous can be the implementation of this command. ( templates should be used for config settings ,,... never for commands, imho)

**Review by jmolmo — DISMISSED** [review]
There are lot of worrying things  in this command and in this modification:
- The original implementation of the command, using a template for specifying a command , is very dangerous.
- The possibility to use the "mon store" to store commands instead config values is very dangerous
- Allow the user to specify commands is dangerous
- We lose the control over the way we are switching lights on/off.

File: src/pybind/mgr/cephadm/tests/test_cephadm.py (hunk 1 above)
Comment (votdev) [reply src/pybind/mgr/cephadm/tests/test_cephadm.py:648]: You're right, its potential a risk, but this PR has not introduced that feature, it only adds some missing behavior. And there is the need to allow the user to modify the command to switch the LEDs on/off.

**Comment by votdev** [comment]
> There are lot of worrying things in this command and in this modification:
> 
>     * The original implementation of the command, using a template for specifying a command , is very dangerous.
> 
>     * The possibility to use the "mon store" to store commands instead config values is very dangerous
> 
>     * Allow the user to specify commands is dangerous
> 
>     * We lose the control over the way we are switching lights on/off.

I will think about how to re-implement this requirement without using mgr store. I think the [approach](https://github.com/SUSE/DeepSea/pull/1474/files#diff-0077b2dee8391093c5ec2d60e593e15467d6cd6e1eed930ae05ca025ea630646R8) i used in a downstream project should be a good solution. Need to find an easy way how to support such a configuration per host.

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -1566,30 +1566,37 @@ def blink_device_light(self, ident_fault: str, on: bool, locs: List[orchestrator
...
+                tpl_name = tpl_name_per_host
+
+            cmd_line = self.template.render(tpl_name, {
                 'on': on,
                 'ident_fault': ident_fault,
                 'dev': dev,
                 'path': path
-            }
-
-            lsmcli_blink_lights_cmd = self.template.render('lsmcli_blink_lights_cmd.j2', j2_ctx)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1582]: render already allows you to do that aready. can you add this to render instead? 

https://github.com/ceph/ceph/blob/8a2f2d82f49df1d5a94e97328f8b04a783d159d5/src/pybind/mgr/cephadm/template.py#L89-L94

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cephadm/tests/test_cephadm.py (hunk 1 above)
Comment (jmolmo) [reply src/pybind/mgr/cephadm/tests/test_cephadm.py:648]: I have realized of the "command in template" problem looking at this modification. This is why i have commented here.

I think that the way to solve it is:

1. Do not use an external template, in this case is just a command, we can store it as an internal variable.

2. Customize by host:
Here is where i see the real problem. In my view we should in any case examine the new platform where we want to use this feature and include the command in our "supported" blinking lights commands set.
We have the gather fact feature to examine the hardware an to know when we need to use one command or another, and thus avoiding the user putting different commands for different hosts in the mon store.

**Comment by jmolmo** [comment]
> > There are lot of worrying things in this command and in this modification:
> > ```
> > * The original implementation of the command, using a template for specifying a command , is very dangerous.
> > 
> > * The possibility to use the "mon store" to store commands instead config values is very dangerous
> > 
> > * Allow the user to specify commands is dangerous
> > 
> > * We lose the control over the way we are switching lights on/off.
> > ```
> 
> I will think about how to re-implement this requirement without using mgr store. I think the [approach](https://github.com/SUSE/DeepSea/pull/1474/files#diff-0077b2dee8391093c5ec2d60e593e15467d6cd6e1eed930ae05ca025ea630646R8) i used in a downstream project should be a good solution. Need to find an easy way how to support such a configuration per host.

Apart of the template  (not needed for this feature at all) )... I think that the core of the problem is not around the hosts , is around support different ways to issue the right command depending of the platform. This requires support this platform, at least in the part about blink lights.

**Comment by jmolmo** [comment]
If you need it ... go ahead ... we know where we need to improve.

File: src/pybind/mgr/cephadm/module.py (hunk 2 above)
Comment (votdev) [reply src/pybind/mgr/cephadm/module.py:1582]: Adding the host as search parameter to the key to access the key-value store has been relocated to the template render method.

**Comment by votdev** [comment]
The intention of this PR is to cleanup the existing code and add the ability to configure the template per host. A documentation how to configure a template has been added, too.
No more improvements are done to the existing code because everything else should be done on the host side, e.g. by overwritting the `lsmcli` command with a bash/Python script that does the needed mapping to an alternative tool in case the HBA is not supported by lsm. All necessary information, e.g. device ID or path or state of the LED, can be extracted by the script from the command arguments.

**Comment by mgfritch** [comment]
https://pulpito.ceph.com/mgfritch-2020-11-14_00:11:13-rados:cephadm-wip-mgfritch-testing-2020-11-13-1122-distro-basic-smithi/

unrelated failure:
https://tracker.ceph.com/issues/48142

**Review by mgfritch — APPROVED** [review]

**Comment by mgfritch** [comment]
> If you need it ... go ahead ... we know where we need to improve.

ping @jmolmo, ok to merge?

**Comment by jmolmo** [comment]
> > If you need it ... go ahead ... we know where we need to improve.
> 
> ping @jmolmo, ok to merge?

I you need it this ... go ahead. But I do not want to approve something with the level of risk that introduces this modification. 

I hope we will have a modification in order to avoid to use the current approach (i mean to use a template to store commands, or use a mon store var to store commands) to define the command used to manage lights. In my opinion this should be changed urgently.

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
@jmolmo + @epuertat : https://tracker.ceph.com/issues/48251

**Comment by epuertat** [comment]
Thank you @sebastian-philipp! Given this PR is cephadm's and might be critical for your downstream, feel free to merge.

**Pull request merged**